test/neo4clj/query_builder_test.clj

```diff
@@ -85,22 +76,43 @@
             "MATCH (G__1:Phone:Fragment) WHERE ((G__1.b = 2) OR (G__1.b = 5)) MATCH (G__123) CREATE (G__1)-[r:TEST_RELATION {a: 1}]->(G__123)" by-combined-rel false
             "MATCH (G__1:Phone:Fragment) WHERE ((G__1.b = 2) OR (G__1.b = 5)) MATCH (G__123) CREATE (G__1)-[r:TEST_RELATION {a: 1}]->(G__123) RETURN r" by-combined-rel true))))))
 
-(t/deftest node-reference
-  (t/testing "Cypher to lookup or reference a already lookded up node"
-    (t/are [expected-cypher-parts known-ref-ids node]
-        (= expected-cypher-parts (sut/node-reference known-ref-ids node))
-      ["(G__123:Person)" nil] #{} {:ref-id "G__123" :labels [:person]}
-      ["(G__123 {firstName: 'Neo'})" nil] #{} {:ref-id "G__123" :props {:first-name "Neo"}}
-      ["(G__123)" "ID(G__123) = 14"] #{} {:ref-id "G__123" :id 14}
-      ["(G__123)" nil] #{"G__123"} {:ref-id "G__123" :labels [:person]}
-      ["(G__123)" nil] #{"G__123"} {:ref-id "G__123" :props {:first-name "Neo"}}
-      ["(G__123)" nil] #{"G__123"} {:ref-id "G__123" :id 14})))
+(t/deftest create-graph-query
+  (t/testing "Cypher to create a graph"
+    (let [graph-map {:lookups [{:ref-id "n1" :labels [:person] :id 4}]
+                     :nodes [{:ref-id "n2" :labels [:company]}]
+                     :rels [{:ref-id "r1" :type :employee :from "n1" :to {:ref-id "n2"}}]
+                     :returns ["n1" {:ref-id "n2"} "r1"]}]
+      (t/are [expected-cypher graph-spec]
+          (= expected-cypher (sut/create-graph-query graph-spec))
+        "MATCH (n1:Person) WHERE ID(n1) = 4 CREATE (n1)-[r1:EMPLOYEE]->(n2:Company) RETURN n1, n2, r1" graph-map
+        "MATCH (n1:Person) WHERE ID(n1) = 4 MATCH (n2:Company) CREATE (n1)-[r1:EMPLOYEE]->(n2) RETURN n1, n2, r1" (dissoc (update-in graph-map [:lookups] concat (:nodes graph-map)) :nodes)
```

Comment:
I would recommend writing the actual `graph-map` used every time over manipulating it with `dissoc` and `update-in`. Yes it is more verbose but it is also easier to read and reason about. Maybe also use a newline in between data sets:
```
"CYPER"
{:some "graph map"}

"NEXT CYPER"
{:new "graph map"
 :with "more keys"}
```

I think the same is true for `lookup-graph-query`.

Reply:
Changed